src/runtime/parcelset/parcel.cpp

```diff
@@ -5,75 +5,247 @@
+        int comptype = action_->get_component_type();
+
+        // decode the local virtual address of the parcel
+        naming::address::address_type lva = data_.addr_.address_;
+        // by convention, a zero address references either the local
+        // runtime support component or one of the AGAS components
+        if (0 == lva)
+        {
+            switch(comptype)
+            {
+            case components::component_runtime_support:
+                lva = hpx::applier::get_applier().get_runtime_support_raw_gid().get_lsb();
+                break;
+
+            case components::component_agas_primary_namespace:
+                lva = client.get_primary_ns_lva();
+                break;
 
-        error_code ec(lightweight);        // ignore all errors
-        std::uint32_t locality_id = hpx::get_locality_id(ec);
-        if (locality_id == naming::invalid_locality_id)
-            locality_id = locality_id_default;
+            case components::component_agas_symbol_namespace:
+                lva = client.get_symbol_ns_lva();
+                break;
 
-        naming::gid_type result = naming::get_gid_from_locality_id(locality_id);
-        result.set_lsb(++id);
-        return result;
+            case components::component_plain_function:
+                break;
+
+            default:
+                HPX_ASSERT(false);
+            }
+        }
+        else if (comptype == components::component_memory)
+        {
+            HPX_ASSERT(naming::refers_to_virtual_memory(data_.dest_));
+            lva = hpx::applier::get_applier().get_memory_raw_gid().get_lsb();
+        }
+
+        // make sure the component_type of the action matches the
+        // component type in the destination address
+        if (HPX_UNLIKELY(!components::types_are_compatible(
+            data_.addr_.type_, comptype)))
+        {
+            std::ostringstream strm;
+            strm << " types are not compatible: destination_type("
+                  << data_.addr_.type_ << ") action_type(" << comptype
+                  << ") parcel ("  << *this << ")";
+            HPX_THROW_EXCEPTION(bad_component_type,
+                "applier::schedule_action",
+                strm.str());
+        }
+
+        return lva;
     }
 
-    void parcel::serialize(serialization::input_archive & ar, unsigned)
+    void parcel::load_schedule(serialization::input_archive & ar,
+        std::size_t num_thread)
     {
-        ar & data_;
-        if(data_.has_source_id_)
+        using hpx::serialization::detail::polymorphic_id_factory;
+
+        ar >> data_;
+        ar >> cont_;
+
+        std::uint32_t id;
+        ar >> id;
+#if !defined(HPX_DEBUG)
+        action_.reset(polymorphic_id_factory::create<actions::base_action>(id));
+#else
+        std::string name;
+        ar >> name;
+        action_.reset(
+            polymorphic_id_factory::create<actions::base_action>(id, &name));
+#endif
```

Comment:
Am 02.09.2016 6:04 nachm. schrieb "Hartmut Kaiser" <notifications@github.com
>:
>
> In src/runtime/parcelset/parcel.cpp:
>
> > - if(data_.has_source_id_) > + using
hpx::serialization::detail::polymorphic_id_factory; > + > + ar >> data_; >
+ ar >> cont_; > + > + std::uint32_t id; > + ar >> id; > +#if
!defined(HPX_DEBUG) > +
action_.reset(polymorphic_id_factory::create<actions::base_action>(id)); >
+#else > + std::string name; > + ar >> name; > + action_.reset( > +
polymorphic_id_factory::create<actions::base_action>(id, &name)); > +#endif
>
> Why is this serialization code duplicated (see parcel::serialize() below)?

We could certainly factor it out to avoid code duplication.

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.